### DISCUSSION TITLE
Best way for dealing with dependencies in workspaces?

### DISCUSSION THREAD

**Discussion by tamis-laan** [dagster-io/dagster#8094] [Q&A]
So I just started with Dagster and I'm setting up the infrastructure to work with my backend. My setup involves containerized services using docker (compose). This also allows for local development.

I have Digit running and the Dagster deamon. I'm using python packages as workspaces, everything works fine. I can manually reload workspaces when changing code (wish digit would monitor for code changes automatically). 

The only thing I have trouble with are dependencies. 

1. I could use a `requirements.txt` file but this would require me to rebuild the dockerfile for each added dependency and then restarting the container. 

2. I could use virtual environments but this would require me to push these environments into my git repository? Or write some kind of CI/CD code to initialize these virtual environments?

I wish I could specify some command that will be run like `pip install -r requirements.txt` inside the workspace that is executed everytime the workspace starts or is reloaded. 

Any way, whats the best for dealing with depencies??

**Comment by smackesey** [comment]
Thanks for the question @tamis-laan. Are you running the daemon and dagit in Docker containers or just your other services?

If you're running the daemon and `dagit` in containers, then you can easily build your dependencies into the image by modifying the `pip install` line in the `Dockerfile` [shown in the docs](https://docs.dagster.io/deployment/guides/docker#deploying-dagster-to-docker).

If you're not running them in containers, then you should just run them inside a virtual env (locally). You'd need to recreate the environment for CI-- not an expert here but you could either run your CI stuff on a Docker image that includes the environment, or maybe use something like `tox` to build the environment for you when the test is run.

**Reply by tamis-laan** [reply]
@smackesey  thanks for the reply

I have seen the example in the docks and this is how I deal with dependencies atm, however this requires me to rebuild the container when adding dependencies. This is not optimal in my opinion.

**Reply by smackesey** [reply]
Hmmm-- well if I understand, your problem is not really dagster-specific, but really just an issue you'll encounter when running _any_ Python process in a Docker container.

But maybe someone more expert than me has a better idea-- @gibsondan?
